File: tools/openmsx-mcp-launcher.js
/**
 * openMSX launcher for mcp-openmsx on Windows.
 *
 * Node/CreateProcess exits openMSX immediately (code 1) in this environment,
 * while ShellExecute (`cmd /c start`) works. mcp-openmsx waits for
 * %TEMP%\openmsx-default\socket.<spawned-pid>, so this wrapper:
 *   1) starts openMSX via ShellExecute
 *   2) finds the new openMSX PID + socket
 *   3) mirrors the socket file under this wrapper's PID
 *   4) stays alive until killed, then terminates openMSX
 */
const { spawn, execSync } = require('child_process');
const fs = require('fs');
const path = require('path');
const os = require('os');

const OPENMSX_EXE = process.env.OPENMSX_REAL_EXECUTABLE
  || 'C:\\Program Files\\openMSX\\openmsx.exe';
const OPENMSX_CWD = path.dirname(OPENMSX_EXE);
const wrapperPid = process.pid;
const args = process.argv.slice(2);
const tempDir = process.env.TEMP || process.env.TMP || os.tmpdir();
const socketDir = path.join(tempDir, 'openmsx-default');

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function listOpenmsxPids() {
  try {
    const out = execSync('tasklist /FI "IMAGENAME eq openmsx.exe" /FO CSV /NH', {
      encoding: 'utf8',
      stdio: ['ignore', 'pipe', 'ignore'],
    });
    const pids = [];
    for (const line of out.split(/\r?\n/)) {
      const m = line.match(/^"openmsx\.exe","(\d+)"/i);
      if (m) pids.push(Number(m[1]));
    }
    return pids;
  } catch {
    return [];
  }
}

function log(msg) {
  // mcp-openmsx reads stderr diagnostics while connecting
  process.stderr.write(`[openmsx-mcp-launcher] ${msg}\n`);
}

async function main() {
  if (!fs.existsSync(OPENMSX_EXE)) {
    process.stderr.write(`Error: openMSX not found: ${OPENMSX_EXE}\n`);
    process.exit(1);
  }

  fs.mkdirSync(socketDir, { recursive: true });
  const before = new Set(listOpenmsxPids());

  // ShellExecute path that actually starts openMSX on this PC
  const child = spawn('cmd.exe', ['/c', 'start', '', OPENMSX_EXE, ...args], {
    cwd: OPENMSX_CWD,
    stdio: 'ignore',
    windowsHide: true,
    detached: true,
  });
  child.unref();
  log(`started via ShellExecute args=[${args.join(' ')}] wrapperPid=${wrapperPid}`);

  let omsxPid = null;
  for (let i = 0; i < 50; i++) {
    await sleep(100);
    const neu = listOpenmsxPids().filter((p) => !before.has(p));
    if (neu.length) {
      omsxPid = neu[neu.length - 1];
      break;
    }
  }
  if (!omsxPid) {
    process.stderr.write('Error: openMSX process did not appear after ShellExecute\n');
    process.exit(1);
  }
  log(`openMSX pid=${omsxPid}`);

  const srcSocket = path.join(socketDir, `socket.${omsxPid}`);
  const dstSocket = path.join(socketDir, `socket.${wrapperPid}`);

  let port = null;
  for (let i = 0; i < 80; i++) {
    if (fs.existsSync(srcSocket)) {
      try {
        port = fs.readFileSync(srcSocket, 'utf8').trim();
        if (port) break;
      } catch {
        // retry
      }
    }
    await sleep(100);
  }
  if (!port) {
    process.stderr.write(`Error: openMSX socket not found: ${srcSocket}\n`);
    try {
      execSync(`taskkill /PID ${omsxPid} /F`, { stdio: 'ignore' });
    } catch { /* ignore */ }
    process.exit(1);
  }

  fs.writeFileSync(dstSocket, `${port}\n`, 'utf8');
  log(`mirrored socket ${srcSocket} -> ${dstSocket} port=${port}`);

  const cleanup = () => {
    try {
      if (fs.existsSync(dstSocket)) fs.unlinkSync(dstSocket);
    } catch { /* ignore */ }
    try {
      execSync(`taskkill /PID ${omsxPid} /F`, { stdio: 'ignore' });
    } catch { /* ignore */ }
  };

  process.on('exit', cleanup);
  process.on('SIGINT', () => {
    cleanup();
    process.exit(0);
  });
  process.on('SIGTERM', () => {
    cleanup();
    process.exit(0);
  });

  // Stay alive while openMSX runs (mcp-openmsx owns this process handle)
  for (;;) {
    await sleep(1000);
    const still = listOpenmsxPids().includes(omsxPid);
    if (!still) {
      log('openMSX exited; launcher ending');
      cleanup();
      process.exit(0);
    }
  }
}

main().catch((err) => {
  process.stderr.write(`Error: ${err && err.stack ? err.stack : err}\n`);
  process.exit(1);
});

File: tools/openmsx-mcp-launcher.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Threading;

namespace OpenMsxMcpLauncher
{
    class Program
    {
        static int Main(string[] args)
        {
            string exe = Environment.GetEnvironmentVariable("OPENMSX_REAL_EXECUTABLE");
            if (string.IsNullOrWhiteSpace(exe))
                exe = @"C:\Program Files\openMSX\openmsx.exe";

            if (!File.Exists(exe))
            {
                Console.Error.WriteLine("Error: openMSX not found: " + exe);
                return 1;
            }

            string tempDir = Environment.GetEnvironmentVariable("TEMP");
            if (string.IsNullOrWhiteSpace(tempDir))
                tempDir = Path.GetTempPath();
            string socketDir = Path.Combine(tempDir, "openmsx-default");
            Directory.CreateDirectory(socketDir);

            int wrapperPid = Process.GetCurrentProcess().Id;
            var before = GetOpenMsxPids();

            // Prefer explorer.exe on a .lnk (supports args). Direct CreateProcess exits 1.
            string launchTarget;
            string tempLnk = Path.Combine(Path.GetTempPath(), "openmsx-mcp-launch-" + wrapperPid + ".lnk");
            try
            {
                Type t = Type.GetTypeFromProgID("WScript.Shell");
                object sh = Activator.CreateInstance(t);
                object shortcut = t.InvokeMember("CreateShortcut", System.Reflection.BindingFlags.InvokeMethod, null, sh, new object[] { tempLnk });
                Type st = shortcut.GetType();
                st.InvokeMember("TargetPath", System.Reflection.BindingFlags.SetProperty, null, shortcut, new object[] { exe });
                st.InvokeMember("WorkingDirectory", System.Reflection.BindingFlags.SetProperty, null, shortcut, new object[] { Path.GetDirectoryName(exe) });
                st.InvokeMember("Arguments", System.Reflection.BindingFlags.SetProperty, null, shortcut, new object[] { QuoteArgs(args) });
                st.InvokeMember("Save", System.Reflection.BindingFlags.InvokeMethod, null, shortcut, null);
                launchTarget = tempLnk;
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine("[openmsx-mcp-launcher] shortcut failed: " + ex.Message + "; launching exe only");
                launchTarget = exe;
            }
            var psi = new ProcessStartInfo();
            psi.FileName = "explorer.exe";
            psi.Arguments = "\"" + launchTarget + "\"";
            psi.UseShellExecute = false;
            Process.Start(psi);
            Console.Error.WriteLine("[openmsx-mcp-launcher] explorer launch wrapperPid=" + wrapperPid + " target=" + launchTarget);

            int omsxPid = 0;
            for (int i = 0; i < 50; i++)
            {
                Thread.Sleep(100);
                foreach (int pid in GetOpenMsxPids())
                {
                    if (!before.Contains(pid))
                    {
                        omsxPid = pid;
                        break;
                    }
                }
                if (omsxPid != 0) break;
            }
            if (omsxPid == 0)
            {
                Console.Error.WriteLine("Error: openMSX process did not appear");
                return 1;
            }
            Console.Error.WriteLine("[openmsx-mcp-launcher] openMSX pid=" + omsxPid);

            string src = Path.Combine(socketDir, "socket." + omsxPid);
            string dst = Path.Combine(socketDir, "socket." + wrapperPid);
            string port = null;
            for (int i = 0; i < 150; i++)
            {
                // Ensure openMSX is still alive while waiting for the control socket
                bool alive = false;
                foreach (int pid in GetOpenMsxPids())
                {
                    if (pid == omsxPid) { alive = true; break; }
                }
                if (!alive)
                {
                    Console.Error.WriteLine("Error: openMSX exited before creating control socket");
                    return 1;
                }
                if (File.Exists(src))
                {
                    try
                    {
                        port = File.ReadAllText(src).Trim();
                        if (!string.IsNullOrEmpty(port)) break;
                    }
                    catch { }
                }
                Thread.Sleep(100);
            }
            if (string.IsNullOrEmpty(port))
            {
                Console.Error.WriteLine("Error: openMSX socket not found: " + src);
                try
                {
                    foreach (var f in Directory.GetFiles(socketDir, "socket.*"))
                        Console.Error.WriteLine("[openmsx-mcp-launcher] seen " + f);
                }
                catch { }
                TryKill(omsxPid);
                return 1;
            }

            File.WriteAllText(dst, port + Environment.NewLine);
            Console.Error.WriteLine("[openmsx-mcp-launcher] mirrored socket port=" + port);

            Console.CancelKeyPress += (s, e) => { e.Cancel = true; Cleanup(dst, omsxPid); Environment.Exit(0); };
            AppDomain.CurrentDomain.ProcessExit += (s, e) => Cleanup(dst, omsxPid);

            while (true)
            {
                Thread.Sleep(1000);
                bool alive = false;
                foreach (int pid in GetOpenMsxPids())
                {
                    if (pid == omsxPid) { alive = true; break; }
                }
                if (!alive)
                {
                    Console.Error.WriteLine("[openmsx-mcp-launcher] openMSX exited");
                    Cleanup(dst, omsxPid);
                    return 0;
                }
            }
        }

        static string QuoteArgs(string[] args)
        {
            if (args == null || args.Length == 0) return "";
            var parts = new string[args.Length];
            for (int i = 0; i < args.Length; i++)
            {
                string a = args[i];
                if (a.IndexOf(' ') >= 0 || a.IndexOf('"') >= 0)
                    parts[i] = "\"" + a.Replace("\"", "\\\"") + "\"";
                else
                    parts[i] = a;
            }
            return string.Join(" ", parts);
        }

        static System.Collections.Generic.HashSet<int> GetOpenMsxPids()
        {
            var set = new System.Collections.Generic.HashSet<int>();
            foreach (var p in Process.GetProcessesByName("openmsx"))
            {
                try { set.Add(p.Id); } catch { }
            }
            return set;
        }

        static void TryKill(int pid)
        {
            try
            {
                var p = Process.GetProcessById(pid);
                p.Kill();
            }
            catch { }
        }

        static void Cleanup(string dst, int omsxPid)
        {
            try { if (File.Exists(dst)) File.Delete(dst); } catch { }
            TryKill(omsxPid);
        }
    }
}
